gateway/src/cron.rs

```diff
@@ -0,0 +1,427 @@
+                Ordering::Less => {}
+                Ordering::Equal => return Err(anyhow!("validators not unique")),
+                Ordering::Greater => return Err(anyhow!("validators not sorted")),
+            };
+        }
+
+        // check top down msgs
+        for i in 1..self.top_down_msgs.len() {
+            match compare_top_down_msg(&self.top_down_msgs[i - 1], &self.top_down_msgs[i])? {
+                Ordering::Less => {}
+                Ordering::Equal => return Err(anyhow!("top down messages not distinct")),
+                Ordering::Greater => return Err(anyhow!("top down messages not sorted")),
+            };
+        }
+
+        let mh_code = Code::Blake2b256;
+        // TODO: to avoid serialization again, maybe we should perform deserialization in the actor
+        // TODO: dispatch call to save gas? The actor dispatching contains the raw serialized data,
+        // TODO: which we dont have to serialize here again
+        Ok(mh_code.digest(&to_vec(self).unwrap()).to_bytes())
+    }
+}
+
+/// Track all the cron checkpoint submissions of an epoch
+#[derive(Serialize_tuple, Deserialize_tuple, PartialEq, Eq, Clone)]
+pub struct CronSubmission {
+    /// All the submitters
+    submitters: TCid<THamt<Address, ()>>,
+    /// The most submitted hash. Using set because there might be a tie
+    most_submitted_hashes: Option<HashSet<HashOutput>>,
+    /// The map to track the max submitted
+    submission_counts: TCid<THamt<HashOutput, u16>>,
+    /// The different cron checkpoints, with cron checkpoint hash as key
+    submissions: TCid<THamt<HashOutput, CronCheckpoint>>,
+}
+
+impl CronSubmission {
+    pub fn new<BS: Blockstore>(store: &BS) -> anyhow::Result<CronSubmission> {
+        Ok(CronSubmission {
+            submitters: TCid::new_hamt(store)?,
+            most_submitted_hashes: None,
+            submission_counts: Default::default(),
+            submissions: TCid::new_hamt(store)?,
+        })
+    }
+
+    /// Submit a cron checkpoint as the submitter. Returns `true` if the submission threshold
+    /// is reached, else `false`.
+    pub fn submit<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        submitter: Address,
+        checkpoint: CronCheckpoint,
+    ) -> anyhow::Result<bool> {
+        // TODO: Add validation of validator set logic so that we know the set of validators is correct
+        let total_validators = checkpoint.validators.validators().len();
+
+        self.update_submitters(store, submitter)?;
+
+        let checkpoint_hash = self.insert_checkpoint(store, checkpoint)?;
+        let most_submitted_count = self.update_submission_count(store, checkpoint_hash)?;
+
+        // use u16 numerator and denominator to avoid floating point calculation and external crate
+        // total validators should be within u16::MAX.
+        if total_validators as u16 * RATIO_NUMERATOR / RATIO_DENOMINATOR > most_submitted_count {
```

Comment:
To compute 2/3 I think we should use the weights of validators so we have a weighted voting